File: src/python/dxpy/bindings/dxanalysis.py
# Copyright (C) 2013-2015 DNAnexus, Inc.
#
# This file is part of dx-toolkit (DNAnexus platform client libraries).
#
#   Licensed under the Apache License, Version 2.0 (the "License"); you may not
#   use this file except in compliance with the License. You may obtain a copy
#   of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#   License for the specific language governing permissions and limitations
#   under the License.

"""
DXAnalysis Handler
++++++++++++++++++

Analyses are DNAnexus entities that capture an instantiation of a
running workflow. They can be created from
:func:`dxpy.bindings.dxworkflow.DXWorkflow.run` or from an
existing analysis ID.

"""

from __future__ import (print_function, unicode_literals)

import time
import dxpy
from dxpy.bindings import (DXObject, )
from dxpy.exceptions import DXJobFailureError

##############
# DXAnalysis #
##############

class DXAnalysis(DXObject):
    '''
    Remote analysis object handler.
    '''

    _class = "analysis"

    def __init__(self, dxid=None):
        self._test_harness_result = None
        DXObject.__init__(self, dxid=dxid)
        self.set_id(dxid)

    def describe(self, fields=None, **kwargs):
        """
        :param fields: dict where the keys are field names that should
            be returned, and values should be set to True (by default,
            all fields are returned)
        :type fields: dict
        :returns: Description of the analysis
        :rtype: dict

        Returns a hash with key-value pairs containing information
        about the analysis

        """
        describe_input = {}
        if fields is not None:
            describe_input['fields'] = fields
        self._desc = dxpy.api.analysis_describe(self._dxid, describe_input, **kwargs)
        return self._desc

    def add_tags(self, tags, **kwargs):
        """
        :param tags: Tags to add to the analysis
        :type tags: list of strings

        Adds each of the specified tags to the analysis. Takes no
        action for tags that are already listed for the analysis.

        """

        dxpy.api.analysis_add_tags(self._dxid, {"tags": tags}, **kwargs)

    def remove_tags(self, tags, **kwargs):
        """
        :param tags: Tags to remove from the analysis
        :type tags: list of strings

        Removes each of the specified tags from the analysis. Takes
        no action for tags that the analysis does not currently have.

        """

        dxpy.api.analysis_remove_tags(self._dxid, {"tags": tags}, **kwargs)

    def set_properties(self, properties, **kwargs):
        """
        :param properties: Property names and values given as key-value pairs of strings
        :type properties: dict

        Given key-value pairs in *properties* for property names and
        values, the properties are set on the analysis for the given
        property names. Any property with a value of :const:`None`
        indicates the property will be deleted.

        .. note:: Any existing properties not mentioned in *properties*
           are not modified by this method.

        """

        dxpy.api.analysis_set_properties(self._dxid, {"properties": properties}, **kwargs)

    def wait_on_done(self, interval=2, timeout=3600*24*7, **kwargs):
        '''
        :param interval: Number of seconds between queries to the analysis's state
        :type interval: integer
        :param timeout: Maximum amount of time to wait, in seconds, until the analysis is done (or at least partially failed)
        :type timeout: integer
        :raises: :exc:`~dxpy.exceptions.DXError` if the timeout is reached before the analysis has finished running, or :exc:`~dxpy.exceptions.DXJobFailureError` if some job in the analysis has failed

        Waits until the analysis has finished running.
        '''

        elapsed = 0
        while True:
            state = self._get_state(**kwargs)
            if state == "done":
                break
            if state in ["failed", "partially_failed"]:
                desc = self.describe(**kwargs)
                err_msg = "Analysis has failed because of {failureReason}: {failureMessage}".format(**desc)
                if desc.get("failureFrom") != None and desc["failureFrom"]["id"] != desc["id"]:
                    err_msg += " (failure from {id})".format(id=desc['failureFrom']['id'])
                raise DXJobFailureError(err_msg)
            if state == "terminated":
                raise DXJobFailureError("Analysis was terminated.")

            if elapsed >= timeout or elapsed < 0:
                raise DXJobFailureError("Reached timeout while waiting for the analysis to finish")

            time.sleep(interval)
            elapsed += interval

    def terminate(self, **kwargs):
        '''
        Terminates the associated analysis.
        '''
        dxpy.api.analysis_terminate(self._dxid, **kwargs)

    def get_output_ref(self, field, index=None, metadata=None):
        '''
        :param field: Output field name of this analysis
        :type field: string
        :param index: If the referenced field is an array, optionally specify an index (starting from 0) to indicate a particular member of the array
        :type index: int
        :param metadata: If the referenced field is of a data object class, a string indicating the metadata that should be read, e.g. "name", "properties.propkey", "details.refgenome"
        :type metadata: string

        Returns a dict containing a valid reference to an output of this analysis.
        '''

        link = {"$dnanexus_link": {"analysis": self._dxid, "field": field}}
        if index is not None:
            link["$dnanexus_link"]["index"] = index
        if metadata is not None:
            link["$dnanexus_link"]["metadata"] = metadata
        return link

    def _get_state(self, **kwargs):
        '''
        :returns: State of the remote object
        :rtype: string

        Queries the API server for the analysis's state.

        Note that this function is shorthand for:

            dxanalysis.describe(**kwargs)["state"]

        '''

        return self.describe(fields=dict(state=True), **kwargs)["state"]

File: src/java/src/main/java/com/dnanexus/DXAnalysis.java
// Copyright (C) 2013-2016 DNAnexus, Inc.
//
// This file is part of dx-toolkit (DNAnexus platform client libraries).
//
//   Licensed under the Apache License, Version 2.0 (the "License"); you may
//   not use this file except in compliance with the License. You may obtain a
//   copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
//   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
//   License for the specific language governing permissions and limitations
//   under the License.

package com.dnanexus;

import java.util.Set;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.collect.Sets;

/**
 * An analysis object (a specific instantiation of a workflow).
 */
public final class DXAnalysis extends DXExecution {

    /**
     * A response from the /analysis-xxxx/terminate route.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class AnalysisTerminateResponse {}

    /**
     * Contains metadata about an analysis. All accessors reflect the state of the analysis at the
     * time that this object was created.
     */
    public final static class Describe extends DXExecution.Describe {

        private final DescribeResponseHash describeOutput;

        // TODO: executable
        // TODO: workflow
        // TODO: stages

        @VisibleForTesting
        Describe(DescribeResponseHash describeOutput, DXEnvironment env) {
            super(describeOutput, env);
            this.describeOutput = describeOutput;
        }

        /**
         * Returns the output of the analysis, deserialized to the specified class, or null if no
         * output hash is available. Note that this field is not guaranteed to be complete until the
         * analysis has finished.
         *
         * <p>
         * A partial output hash is available as soon as the first stage finishes. As each stage
         * finishes, its corresponding outputs will become visible.
         * </p>
         *
         * @param outputClass class to deserialize to
         *
         * @return output object or null
         */
        @Override
        public <T> T getOutput(Class<T> outputClass) {
            // We don't do anything different here except for providing the analysis-specific
            // caveats in the docstring above.
            return super.getOutput(outputClass);
        }

        /**
         * Returns the state of the analysis.
         *
         * @return analysis state
         */
        public AnalysisState getState() {
            return describeOutput.state;
        }
    }

    /**
     * Deserialized output from the /analysis-xxxx/describe route. Not directly accessible by users
     * (see Describe instead).
     */
    @VisibleForTesting
    @JsonIgnoreProperties(ignoreUnknown = true)
    final static class DescribeResponseHash extends DXExecution.DescribeResponseHash {
        @JsonProperty
        private AnalysisState state;
    }

    private static final Set<AnalysisState> unsuccessfulAnalysisStates = Sets.immutableEnumSet(
            AnalysisState.FAILED, AnalysisState.TERMINATED);

    /**
     * Returns a {@code DXAnalysis} representing the specified analysis.
     *
     * @param analysisId Analysis ID, of the form {@code "analysis-xxxx"}
     *
     * @return a {@code DXAnalysis}
     *
     * @throws NullPointerException If {@code analysisId} is null
     */
    public static DXAnalysis getInstance(String analysisId) {
        return new DXAnalysis(analysisId);
    }

    /**
     * Returns a {@code DXAnalysis} representing the specified analysis using the specified
     * environment, with the specified cached describe output.
     *
     * <p>
     * This method is for use exclusively by bindings to the "find" routes when describe hashes are
     * returned with the find output.
     * </p>
     *
     * @throws NullPointerException If any argument is null
     */
    static DXAnalysis getInstanceWithCachedDescribe(String jobId, DXEnvironment env,
            JsonNode describe) {
        return new DXAnalysis(jobId, Preconditions.checkNotNull(env, "env may not be null"),
                Preconditions.checkNotNull(describe, "describe may not be null"));
    }

    /**
     * Returns a {@code DXAnalysis} representing the specified analysis using the specified
     * environment.
     *
     * @param analysisId Analysis ID, of the form {@code "analysis-xxxx"}
     * @param env environment to use for making subsequent API calls
     *
     * @return a {@code DXAnalysis}
     *
     * @throws NullPointerException If {@code analysisId} or {@code env} is null
     */
    public static DXAnalysis getInstanceWithEnvironment(String analysisId, DXEnvironment env) {
        return new DXAnalysis(analysisId, Preconditions.checkNotNull(env, "env may not be null"));
    }

    private DXAnalysis(String analysisId) {
        super(analysisId, "analysis", null);
    }

    private DXAnalysis(String analysisId, DXEnvironment env) {
        super(analysisId, "analysis", env);
    }

    private DXAnalysis(String analysisId, DXEnvironment env, JsonNode cachedDescribe) {
        super(analysisId, "analysis", env, cachedDescribe);
    }

    @Override
    public Describe describe() {
        return describeImpl(MAPPER.createObjectNode());
    }

    private Describe describeImpl(JsonNode describeInput) {
        return new Describe(DXAPI.analysisDescribe(this.getId(), describeInput,
                DescribeResponseHash.class), this.env);
    }

    @Override
    public Describe getCachedDescribe() {
        this.checkCachedDescribeAvailable();
        return new Describe(
                DXJSON.safeTreeToValue(this.cachedDescribe, DescribeResponseHash.class), this.env);
    }

    @Override
    public <T> T getOutput(Class<T> outputClass) throws IllegalStateException {
        // {fields: {output: true, state: true}}
        Describe d =
                describeImpl(DXJSON
                        .getObjectBuilder()
                        .put("fields",
                                DXJSON.getObjectBuilder().put("output", true).put("state", true)
                                        .build()).build());
        if (d.getState() != AnalysisState.DONE) {
            throw new IllegalStateException(
                    "Expected analysis to be in state DONE, but it is in state " + d.getState());
        }
        return d.getOutput(outputClass);
    }

    @Override
    public void terminate() {
        DXAPI.analysisTerminate(this.getId(), AnalysisTerminateResponse.class);
    }

    /**
     * Waits until the analysis has successfully completed and is in the DONE state.
     *
     * @return the same DXAnalysis object
     *
     * @throws IllegalStateException if the analysis reaches the FAILED or TERMINATED state
     */
    @Override
    public DXAnalysis waitUntilDone() throws IllegalStateException {
        AnalysisState analysisState = this.describe().getState();
        while (analysisState != AnalysisState.DONE) {
            if (unsuccessfulAnalysisStates.contains(analysisState)) {
                throw new IllegalStateException(this.getId() + " is in unsuccessful state "
                        + analysisState.toString());
            }
            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            analysisState = this.describe().getState();
        }
        return this;
    }

}
